### DISCUSSION TITLE
How to receive an OAuth 2.0 token?

### DISCUSSION THREAD

**Discussion by Coooz** [nextgenhealthcare/connect#6113] [Q&A]
Hi,

I am trying to set up an HTTP Sender destination in order to obtain an OAuth 2.0 token, as described in https://open.dips.no/docs/authentication/dfs. So far I have only succeeded in getting an "invalid request" returned from the server, but I cannot put my finger on what I do wrong. This is my destination:
![image](https://github.com/nextgenhealthcare/connect/assets/88899312/dc78dd71-e408-4cb5-8427-7f607406bdec)

And this is what I send:
![image](https://github.com/nextgenhealthcare/connect/assets/88899312/9c45e4e2-87bc-4add-adcc-f86df9c96704)

I _am_ getting a token if I do the exercise in Postman, as described on the same website under "Using Postman".

What do I need to adjust to my destination in order to receive a token?

**Comment by pacmano1** [comment]
if your postman request results in a successful response, send it to a dumb echo server that shows you exactly what postman sends.

Then send the same thing with mirth, and compare the output.   it is almost always obvious how your mirth request is wrong.

A dumb echo server can be started from the command line if you have node.ja installed with:

npx http-echo-server

**Comment by Coooz** [comment]
Hi @pacmano1 ,

Thanks for replying. I am not entirely sure how to send the request to a dumb echo server, which I have running now. My command window shows `[server] event: listening (port: 3000)`.
I am not very familiar with the subject. What I have tried in Postman is making a duplicate of the POST that gives me the token, but setting the Callback URL to http://127.0.0.1:3000 and http://0.0.0.0:3000. This does not work, which probably will be obvious to anyone with some more knowledge on the subject. So... how do I do that?

**Reply by pacmano1** [reply]
You said you are not getting a token calling the endpoint with Mirth.   Change the endpoint to your local echo server.   The callback URL is irrelevant here.   

If what you are trying to accomplish is sorta like "sign in with google" to some other system, that is a different thing.

**Reply by Coooz** [reply]
No, I am not trying to do anything like "sign in with google". I ultimately need to retrieve resources from a FHIR server, which as such is not that hard. I'm only struggling with the authentication part.

Anyway, making _some_ progress here (although not much)...
This is what my dumb echo server shows:

```
--> POST / HTTP/1.1
--> Content-Type: application/x-www-form-urlencoded
--> Authorization: Basic cG9zdG1hbjpwb3N0bWFu
--> User-Agent: PostmanRuntime/7.36.3
--> Accept: */*
--> Cache-Control: no-cache
--> Postman-Token: 90207531-b05d-4a4b-a04d-f2b5e949c58b
--> Host: localhost:3000
--> Accept-Encoding: gzip, deflate, br
--> Connection: keep-alive
--> Content-Length: 136
-->
--> grant_type=authorization_code&code=kgLf5PABNyYl6xs2Pt86ekuEIjL7i1g22yfEeHO915Q&redirect_uri=https%3A%2F%2Foauth.pstmn.io%2Fv1%2Fcallback
```

I am uncertain about:
- "cG9zdG1hbjpwb3N0bWFu" in the Authorization line - where does it come from? I can only give "Basic" here - I cannot see this value in Postman either.
- The Postman-Token: I can provide my own guid here of course, but is that the right thing to do? In Postman it says "(calculated when request is sent)".
- Host: here it also says in Postman, "(calculated when request is sent)" - so what do I do with it in Mirth Connect?
-  "&code=kgLf5PABNyYl6xs2Pt86ekuEIjL7i1g22yfEeHO915Q&redirect_uri=https%3A%2F%2Foauth.pstmn.io%2Fv1%2Fcallback" in the last line. How do I get the code, and must I provide a redirect_uri in Mirth Connect? I suppose not.

In addition I get the idea that I am thinking way too complicated:
A). ggVert on https://github.com/nextgenhealthcare/connect/discussions/4775 is doing something similar, but his HTTP Sender has far less headers and parameters and he at least gets a 200 status in return. I've tried to remodel my destination like ggVert's but to no avail;
B). The open DIPS website (https://open.dips.no/docs/authentication/dfs) clearly states that "the following information can be used to obtain access tokens" (not on their "Using Postman" page though) and lists only six parameters:
![image](https://github.com/nextgenhealthcare/connect/assets/88899312/e81b072c-a3c1-4847-95b6-7b6cee13faf0)
However when I implement that in Mirth Connect as follows, I once more get an "invalid_request" returned.
![image](https://github.com/nextgenhealthcare/connect/assets/88899312/158f1fea-f883-4b42-8abd-e7d585b2fca3)
It goes without saying that any additional help will be _very_ much appreciated.

**Comment by pacmano1** [comment]
![image](https://github.com/nextgenhealthcare/connect/assets/44065187/3c6701aa-f565-4083-96a9-d9e7f2c26361)

**Reply by pacmano1** [reply]
If I understand their documents correctly, you have to stand up an endpoint that redirects the user to login (sending the correct headers to their side) which in turn posts back to you.   Note the request method will be different so you branch on that on the http listener to then go get a token.     This normally means you need a real certifcate on your endpoint.   That is where the callback URL applies.   I may be misreading their documentation of course, so maybe others will chime in.

I don't know how Postman sorta hacks its way through the above process.
